Импортируем модули для работы с данными

In [384]:
import pandas as pd

Считаем датасет

In [385]:
df = pd.read_csv('train.csv')

Выведем 5 первых строчек из датасета

In [386]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Просмотр информации по данным, из этой информации мы можем узнать тип данных колонки, их количество, а так же количестов не нулевых данных по колонкам

In [387]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Вывод общего количества строк и колонок

In [388]:
df.shape

(891, 12)

Выведем количество, среднее, стандартное отклонение, перцентали,  минимальные и максимальные значения по колонкам

In [389]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Подсчитаем процент выживаемости у каждого класса пассажиров

In [390]:
total_passenger = df["PassengerId"].count(); #Общее количество

df.groupby("Pclass")["Survived"].sum() * 100 / total_passenger #Процент выживаемости

Pclass
1    15.263749
2     9.764310
3    13.355780
Name: Survived, dtype: float64

 ## Выведем самое популярное мужское и самое популярное женское имя на корабле

Объявим функцию по поиску и извлечению женского имени с использованием регулярного выражения (в функции используется второй поиск для тех случаев когда замужняя девушка имеет имя мужа и в круглых скобках своё имя, например - Futrelle, Mrs. Jacques Heath (Lily May Peel))

In [391]:
def parse_first_name_female(name):
    first = name.str.extract(r"Mrs\.\s+[^(]*\((\w+)", expand=False)
    first.loc[first.isna()] = name.str.extract(r"\.\s+(\w+)", expand=False)
    return first

Объявим функцию по поиску и извлечению мужского имени 

In [392]:
def parse_first_name_male(name):
    first = name.str.extract(r"(Mr\.|Dr\.|Rev\.|Master\.|Don\.|Major\.|Sir\.|Col\.|Capt\.|Jonkheer\.) ([a-zA-Z]+)", expand=False)[1]
    return first

In [393]:
male_names = parse_first_name_male(df.loc[df["Sex"] == "male", "Name"])
male_names.value_counts(ascending=False).head(1)

1
William    35
Name: count, dtype: int64

In [394]:
femaly_names = parse_first_name_female(df.loc[df["Sex"] == "female", "Name"])
femaly_names.value_counts().head(1)

Name
Anna    15
Name: count, dtype: int64

По итогу самое частое имя среди мужчин - William, а среди женщин - Anna.

## Выведем самое популярное мужское и самое популярное женское имя на корабле в каждом классе

In [395]:
df_male = df[df["Sex"] == "male"]
df_male["First Name"] = male_names

/var/folders/l5/6n4jy99s0sb0vw4z75hwvsxh0000gn/T/ipykernel_23455/1940479345.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_male["First Name"] = male_names


In [396]:
df_male.head()

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0            1         0       3         Braund, Mr. Owen Harris  male  22.0   
4            5         0       3        Allen, Mr. William Henry  male  35.0   
5            6         0       3                Moran, Mr. James  male   NaN   
6            7         0       1         McCarthy, Mr. Timothy J  male  54.0   
7            8         0       3  Palsson, Master. Gosta Leonard  male   2.0   

   SibSp  Parch     Ticket     Fare Cabin Embarked First Name  
0      1      0  A/5 21171   7.2500   NaN        S       Owen  
4      0      0     373450   8.0500   NaN        S    William  
5      0      0     330877   8.4583   NaN        Q      James  
6      0      0      17463  51.8625   E46        S    Timothy  
7      3      1     349909  21.0750   NaN        S      Gosta

Проверим мужские имена на пропуски в данных

In [397]:
df_male[df_male["First Name"].isna()]

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked, First Name]
Index: []

In [398]:
df_female = df[df["Sex"] == "female"]
df_female["First Name"] = femaly_names

/var/folders/l5/6n4jy99s0sb0vw4z75hwvsxh0000gn/T/ipykernel_23455/2444198223.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_female["First Name"] = femaly_names


In [399]:
df_female.head()

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked First Name  
1      0          PC 17599  71.2833   C85        C   Florence  
2      0  STON/O2. 3101282   7.9250   NaN        S      Laina  
3      0            113803  53.1000  C123        S       Lily  
8      2            347742  11.1333   NaN        S  Elisabeth  
9      0            237736  30.0708   NaN        C      Adele

Проверим женские имена на пропуски в данных

In [400]:
df_female[df_female["First Name"].isna()]

PassengerId  Survived  Pclass  \
556          557         1       1   
564          565         0       3   

                                                  Name     Sex   Age  SibSp  \
556  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female  48.0      1   
564                     Meanwell, Miss. (Marion Ogden)  female   NaN      0   

     Parch             Ticket   Fare Cabin Embarked First Name  
556      0              11755  39.60   A16        C        NaN  
564      0  SOTON/O.Q. 392087   8.05   NaN        S        NaN

Заполним в ручную пропуски данных именами женщин

In [401]:
df_female.loc[df_female["PassengerId"] == 557, ["First Name"]] = str("Lucille")

In [402]:
df_female.loc[df_female["PassengerId"] == 565, ["First Name"]] = str("Marion")

In [403]:
df_female.loc[df_female["PassengerId"] == 557]

PassengerId  Survived  Pclass  \
556          557         1       1   

                                                  Name     Sex   Age  SibSp  \
556  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female  48.0      1   

     Parch Ticket  Fare Cabin Embarked First Name  
556      0  11755  39.6   A16        C    Lucille

In [404]:
df_female.loc[df_female["PassengerId"] == 565]

PassengerId  Survived  Pclass                            Name     Sex  \
564          565         0       3  Meanwell, Miss. (Marion Ogden)  female   

     Age  SibSp  Parch             Ticket  Fare Cabin Embarked First Name  
564  NaN      0      0  SOTON/O.Q. 392087  8.05   NaN        S     Marion

Выполним группировку

In [405]:
pclass_name_female = df_female["First Name"].groupby(df_female["Pclass"]).value_counts()
pclass_name_female.groupby(level=0).nlargest(1).reset_index(level=0, drop=True)

Pclass  First Name
1       Elizabeth     5
2       Elizabeth     5
3       Anna          9
Name: count, dtype: int64

In [406]:
pclass_name_male = df_male["First Name"].groupby(df_male["Pclass"]).value_counts()
pclass_name_male.groupby(level=0).nlargest(1).reset_index(level=0, drop=True)

Pclass  First Name
1       William       11
2       William        9
3       William       15
Name: count, dtype: int64

По итогу: среди женских имен в первом классе и во втором популярное имя - Elizabeth, в третьем классе - Anna; среди мужчин во всех трёх классов популярное имя - William

## Выведем часть таблицы с пассажирами, возраст которых больше 44 лет

In [407]:
df[df["Age"] > 44]

PassengerId  Survived  Pclass  \
6              7         0       1   
11            12         1       1   
15            16         1       2   
33            34         0       2   
52            53         1       1   
..           ...       ...     ...   
857          858         1       1   
862          863         1       1   
871          872         1       1   
873          874         0       3   
879          880         1       1   

                                                  Name     Sex   Age  SibSp  \
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
33                               Wheadon, Mr. Edward H    male  66.0      0   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
..                                                 ...     ...   ...    ...   
857                             Daly, Mr. Peter Denis     male  51.0      0   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
873                        Vander Cruyssen, Mr. Victor    male  47.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   

     Parch      Ticket     Fare Cabin Embarked  
6        0       17463  51.8625   E46        S  
11       0      113783  26.5500  C103        S  
15       0      248706  16.0000   NaN        S  
33       0  C.A. 24579  10.5000   NaN        S  
52       0    PC 17572  76.7292   D33        C  
..     ...         ...      ...   ...      ...  
857      0      113055  26.5500   E17        S  
862      0       17466  25.9292   D17        S  
871      1       11751  52.5542   D35        S  
873      0      345765   9.0000   NaN        S  
879      1       11767  83.1583   C50        C  

[115 rows x 12 columns]

## Выведем часть таблицы с пассажирами, возраст которых меньше 44 лет и которые мужского пола

In [408]:
df[(df["Age"] < 44) & (df["Sex"] == 'male')]

PassengerId  Survived  Pclass                            Name   Sex  \
0              1         0       3         Braund, Mr. Owen Harris  male   
4              5         0       3        Allen, Mr. William Henry  male   
7              8         0       3  Palsson, Master. Gosta Leonard  male   
12            13         0       3  Saundercock, Mr. William Henry  male   
13            14         0       3     Andersson, Mr. Anders Johan  male   
..           ...       ...     ...                             ...   ...   
883          884         0       2   Banfield, Mr. Frederick James  male   
884          885         0       3          Sutehall, Mr. Henry Jr  male   
886          887         0       2           Montvila, Rev. Juozas  male   
889          890         1       1           Behr, Mr. Karl Howell  male   
890          891         0       3             Dooley, Mr. Patrick  male   

      Age  SibSp  Parch            Ticket    Fare Cabin Embarked  
0    22.0      1      0         A/5 21171   7.250   NaN        S  
4    35.0      0      0            373450   8.050   NaN        S  
7     2.0      3      1            349909  21.075   NaN        S  
12   20.0      0      0         A/5. 2151   8.050   NaN        S  
13   39.0      1      5            347082  31.275   NaN        S  
..    ...    ...    ...               ...     ...   ...      ...  
883  28.0      0      0  C.A./SOTON 34068  10.500   NaN        S  
884  25.0      0      0   SOTON/OQ 392076   7.050   NaN        S  
886  27.0      0      0            211536  13.000   NaN        S  
889  26.0      0      0            111369  30.000  C148        C  
890  32.0      0      0            370376   7.750   NaN        Q  

[368 rows x 12 columns]

 ## Выведем количества n-местных кабин (в которых было 2, 3, 4, ... человека)

Заполним пустые поля в номере кабины значением - "Cabin missing"

In [409]:
df["Cabin"] = df["Cabin"].apply(lambda x: "Cabin missing" if pd.isnull(x) else x)

На один билет было зарегистрировано несколько человек, например - 347082

In [410]:
df["Ticket"].value_counts()

Ticket
347082      7
CA. 2343    7
1601        7
3101295     6
CA 2144     6
           ..
9234        1
19988       1
2693        1
PC 17612    1
370376      1
Name: count, Length: 681, dtype: int64

In [411]:
df[df["Ticket"] == '347082']

PassengerId  Survived  Pclass  \
13            14         0       3   
119          120         0       3   
541          542         0       3   
542          543         0       3   
610          611         0       3   
813          814         0       3   
850          851         0       3   

                                                  Name     Sex   Age  SibSp  \
13                         Andersson, Mr. Anders Johan    male  39.0      1   
119                  Andersson, Miss. Ellis Anna Maria  female   2.0      4   
541               Andersson, Miss. Ingeborg Constanzia  female   9.0      4   
542                  Andersson, Miss. Sigrid Elisabeth  female  11.0      4   
610  Andersson, Mrs. Anders Johan (Alfrida Konstant...  female  39.0      1   
813                 Andersson, Miss. Ebba Iris Alfrida  female   6.0      4   
850            Andersson, Master. Sigvard Harald Elias    male   4.0      4   

     Parch  Ticket    Fare          Cabin Embarked  
13       5  347082  31.275  Cabin missing        S  
119      2  347082  31.275  Cabin missing        S  
541      2  347082  31.275  Cabin missing        S  
542      2  347082  31.275  Cabin missing        S  
610      5  347082  31.275  Cabin missing        S  
813      2  347082  31.275  Cabin missing        S  
850      2  347082  31.275  Cabin missing        S

Это были пассажиры, которые были семьей, к сожалению в примере никто не выжил

Зададим две функции по определению вместимости кабины и номера класса

In [412]:
def get_cabin_capacity(cabin_obj):
    for key, value in cabin_obj.to_dict().items():
        if(value > 1):
            print(f'{value} - местная кабина № {key}, находится в {get_class_number_by_cabin(key)} классе.')

        
def get_class_number_by_cabin(cabin):
    return df.loc[df["Cabin"] == cabin, "Pclass"].iloc[0]


get_cabin_capacity(df["Cabin"].value_counts())

687 - местная кабина № Cabin missing, находится в 3 классе.
4 - местная кабина № C23 C25 C27, находится в 1 классе.
4 - местная кабина № G6, находится в 3 классе.
4 - местная кабина № B96 B98, находится в 1 классе.
3 - местная кабина № C22 C26, находится в 1 классе.
3 - местная кабина № D, находится в 2 классе.
3 - местная кабина № F33, находится в 2 классе.
3 - местная кабина № E101, находится в 2 классе.
3 - местная кабина № F2, находится в 2 классе.
2 - местная кабина № B20, находится в 1 классе.
2 - местная кабина № E67, находится в 1 классе.
2 - местная кабина № C125, находится в 1 классе.
2 - местная кабина № E24, находится в 1 классе.
2 - местная кабина № B49, находится в 1 классе.
2 - местная кабина № B77, находится в 1 классе.
2 - местная кабина № D35, находится в 1 классе.
2 - местная кабина № C78, находится в 1 классе.
2 - местная кабина № C93, находится в 1 классе.
2 - местная кабина № C65, находится в 1 классе.
2 - местная кабина № B57 B59 B63 B66, находится в 1 классе.
2 

По итогу: в основном двух, трёх и четырех местные кабины располагались в первом и во втором классе. Для 687 пассажиров не был определен номер кабины в третьем классе, возможно пассажиры находились в холле и/или каюты не были достроены.